text/0000-stabilize-arbitrary-self-types.md

```diff
@@ -227,6 +252,22 @@ A previous PR based on the `Deref` alternative has been proposed before https://
     This fails, because `T: Receiver<Target=T>` generally does not hold.
     An alternative would be to lift the associated type into a generic type parameter of the `Receiver` trait, that would allow adding a blanket `impl Receiver<T> for T` without overlap.
 
+- Should this work for (statically resolved) trait method calls? e.g.
+  ```rs
+  use std::ops::Receiver;
+  struct CustomPtr<T>(T);
+  impl<T> Receiver for CustomPtr<T> {
+    type Target = T;
+  }
+  trait Foo {
+     fn bar(self: CustomPtr<Self>);
+  }
+  struct Baz;
+  impl Foo for Baz {
+    fn bar(self: CustomPtr<Foo>) {}
+  }
+  ```
```

Comment:
I am not really sure what you are trying to do.

Is it that one can write the trait name instead of `Self`?

Or are trying something with trait objects? Something like `fn bar(self: CustomPtr<impl Foo>)`?

If it is about allowing the trait name instead of `Self`: I don't like it because it seems confusing. It implies that it can also be another type implementing the trait, while it only can (?) be `Self` (`Baz` in this case). It would be fine to substitute `Self` with `Baz`.

Reply:
Yep, sorry I meant simply:

```
fn bar(self: CustomPtr<Self>) {}
```

Duly adjusted.

In short, I just want to confirm that we expect "arbitrary self types" to work for:
* Methods belonging directly to a type (`impl SomeType { fn method(self: CustomPtr<T>); }`)
* and also, methods in a trait implementation

Such trait method calls do [work with `Rc`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=026195d66879048ef2c6357f35031e2c) so I assume we _do_ want to support them.

Reply:
Ah yes. In my understanding having them in trait methods was one of the main use cases, because then the trait can rely on additional guarantees.